src/pointer.rs

```diff
@@ -0,0 +1,88 @@
+//! Abstractions over raw pointers.
+
+mod ptr;
+
+pub use maybe_aligned::MaybeAligned;
+pub use ptr::{invariant, Ptr};
+
+/// A shorthand for a maybe-valid, maybe-aligned reference. Used as the argument
+/// to [`TryFromBytes::is_bit_valid`][crate::TryFromBytes::is_bit_valid].
+pub type MaybeValid<'a, T> =
+    Ptr<'a, T, { invariant::Shared }, { invariant::AnyAlignment }, { invariant::AsInitialized }>;
+
+/// A semi-user-facing wrapper type representing a maybe-aligned reference, for
+/// use in [`TryFromBytes::is_bit_valid`][crate::TryFromBytes::is_bit_valid].
+mod maybe_aligned {
+    use super::{ptr::invariant, Ptr};
+    use crate::Unaligned;
+    use core::fmt::{Debug, Formatter};
+
+    /// A reference to a validly-initialized `T` of lifetime `'a` that might
+    /// *not* be validly aligned for `T`.
+    //  NOTE: Although this type is called `MaybeAligned`, that name is only
+    //  appropriate in its sole place of use: `TryFromBytes::is_bit_valid`.
+    //  There, the preconditions on the function's argument ensure that
+    #[doc(hidden)]
+    #[allow(unreachable_pub)] // false positive?
+    pub struct MaybeAligned<'a, T: ?Sized> {
+        ptr: Ptr<'a, T, { invariant::Shared }, { invariant::AnyAlignment }, { invariant::Valid }>,
+    }
```

Comment:
Why can't this just be a type alias like `MaybeValid`?

Reply:
Gotcha. Would we ever want to expose other `Ptr` functionality to callers? I worry that doing this sends us down a road in which we'll end up duplicating methods from `Ptr`.

Alternatively, is there no way for us to coax Rust into emitting better error messages while still using an alias?

Reply:
Like this:
```rust
macro_rules! define_invariants {
    ($($(#[$set_attr:meta])* $set:ident {
        $( $(#[$elem_attr:meta])* $elem:ident,)*
    })*) => {
        mod sealed {
            $(  
                #[allow(non_snake_case)]
                pub(super) mod $set {
                    pub trait Sealed {}
                    
                    $(impl Sealed for super::super::$elem {})*
                }
            )*
        }
        $(
            $(#[$set_attr])*
            pub trait $set: sealed::$set::Sealed {}

            $(
                $(#[$elem_attr])*
                pub enum $elem {}
                
                impl $set for $elem {}
            )*
        )*
    }
}

pub mod invariants {
    define_invariants! {
        /// The aliasing invariant of a [`Ptr`][super::Ptr].
        Aliasing {
            /// No aliasing invariant.
            AnyAliasing,
            
            /// The `Ptr<'a, T>` adheres to the aliasing rules of a `&'a T`.
            ///
            /// The referent of a shared-aliased `Ptr` may be concurrently
            /// referenced by any number of shared-aliased `Ptr` or `&T`
            /// references, and may not be concurrently referenced by any
            /// exclusively-aliased `Ptr`s or `&mut T` references.
            ///
            /// Unsafe code must assume that safe code will assume that the
            /// referent of such a `Ptr` will not be mutated, except through
            /// interior [`UnsafeCell`]s (if any).
            ///
            /// [`UnsafeCell`]: core::cell::UnsafeCell
            Shared,
            
            /// The `Ptr<'a, T>` adheres to the aliasing rules of a `&'a mut T`.
            ///
            /// The referent of an exclusively-aliased `Ptr` may not be
            /// concurrently referenced by any other `Ptr`s or references.
            /// Owners of an exclusively-aliased `Ptr` may safely (contigent on
            /// upholding the referent's original validity invariants) mutate
            /// its referent.
            Exclusive,
        }

        /// The alignment invariant of a [`Ptr`][super::Ptr].
        Alignment {
            /// The referent is not necessarily aligned.
            AnyAlignment,
            /// THe referent is aligned.
            Aligned,
        }

        /// The validity invariant of a [`Ptr`][super::Ptr].
        Validity {
            /// The referent is not necessarily initialized.
            AnyValidity,
            
            /// The byte ranges initialized in `T` are also initialized in the
            /// referent.
            ///
            /// Formally: uninitialized bytes may only be present in `Ptr<T>`'s
            /// referent where it is possible for them to be present in `T`.
            /// This is a dynamic property: if, at a particular byte offset, a
            /// valid enum discriminant is set, the subsequent bytes may only
            /// have uninitialized bytes as specificed by the corresponding
            /// enum.
            ///
            /// Formally, given `len = size_of_val_raw(ptr)`, at every byte
            /// offset, `b`, in the range `[0, len)`:
            /// - If, in all instances `t: T` of length `len`, the byte at
            ///   offset `b` in `t` is initialized, then the byte at offset `b`
            ///   within `*ptr` must be initialized.
            /// - Let `c` be the contents of the byte range `[0, b)` in `*ptr`.
            ///   Let `S` be the subset of valid instances of `T` of length
            ///   `len` which contain `c` in the offset range `[0, b)`. If, for
            ///   all instances of `t: T` in `S`, the byte at offset `b` in `t`
            ///   is initialized, then the byte at offset `b` in `*ptr` must be
            ///   initialized.
            ///
            ///   Pragmatically, this means that if `*ptr` is guaranteed to
            ///   contain an enum type at a particular offset, and the enum
            ///   discriminant stored in `*ptr` corresponds to a valid variant
            ///   of that enum type, then it is guaranteed that the appropriate
            ///   bytes of `*ptr` are initialized as defined by that variant's
            ///   bit validity (although note that the variant may contain
            ///   another enum type, in which case the same rules apply
            ///   depending on the state of its discriminant, and so on
            ///   recursively).
            AsInitialized,
            
            /// The referent is bit-valid for `T`.
            Valid,
        }
    }
}
```

Reply:
I've implemented the above strategy, and made `MaybeAligned` merely a alias of `Ptr`.